[quantization] Replacing the skip_list with white_list in the qconfig propagation

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27183 [quantization] Replacing the skip_list with white_list in the qconfig propagation**

Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

Commit:
[quantization] Replacing the skip_list with white_list in the qconfig propagation

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -32,32 +33,22 @@ def _propagate_qconfig_helper(module, qconfig_dict, skip_list=None,
     # TODO: Add test
-    if len(module._modules) == 0 and type(module) == nn.Sequential:
-        module.qconfig = None
+    if white_list is None:
```
Why not have white_list= DEFAULT_QCONFIG_PROPAGATE_WHITE_LIST as a default?

zafartahirov replied:
Because if the `DEFAULT_QCONFIG_PROPAGATE_WHITE_LIST` is mutable, that will create problems

raghuramank100 commented on torch/quantization/quantize.py:
```diff
@@ -32,32 +33,22 @@ def _propagate_qconfig_helper(module, qconfig_dict, skip_list=None,
 
-# TODO(jerryzh): expose skip_list
+# TODO(jerryzh): expose white_list
 def propagate_qconfig_(module, qconfig_dict=None):
```
We should expose the whitelist as part of the top-level API (prepare/prepare_qat/quantize/quantize_qat).

zafartahirov replied:
There is a TODO that jerry left. It might be part of a different PR.

raghuramank100 commented:
Can we have tests to verify the new functionality. The test would have a module with a custom leaf node.
For example:
```python
 class customModule(torch.nn.Module):
  def __init__(self):
   ....

  def forward(self, x):
       return x


 class testModel(nn.Module):
    def __init__(self):
       super(testModel,self).__init__()
       self.conv1 = nn.Conv2d(...)
       self.myMod = customModule()
def forward(self,x):
   x= self.quant(x)
   x = self.conv1(x)
   x = self.dequant(x)
   x  = self.customModule(x)
   return(x)
```
```
myModel = testModel()
myModel.qconfig = default_qconfig
# rest of prepare/convert pipeline
```
- works with the change, would have failed previously

raghuramank100 reviewed (changes requested):
Please see comments, thanks

zafartahirov commented:
> Can we have tests to verify the new functionality. The test would have a module with a custom leaf node.
> For example:
> 
> ```python
>  class customModule(torch.nn.Module):
>   def __init__(self):
>    ....
> 
>   def forward(self, x):
>        return x
> 
> 
>  class testModel(nn.Module):
>     def __init__(self):
>        super(testModel,self).__init__()
>        self.conv1 = nn.Conv2d(...)
>        self.myMod = customModule()
> def forward(self,x):
>    x= self.quant(x)
>    x = self.conv1(x)
>    x = self.dequant(x)
>    x  = self.customModule(x)
>    return(x)
> ```
> 
> ```
> myModel = testModel()
> myModel.qconfig = default_qconfig
> # rest of prepare/convert pipeline
> ```
> 
> * works with the change, would have failed previously

Can you, please, elaborate how would this example fail with the `skip_list` implementation? I want to make sure I understand the unittest well.

raghuramank100 commented:
> Can you, please, elaborate how would this example fail with the `skip_list` implementation? I want to make sure I understand the unittest well.

With the current implementation, since we are based on skiplist, we would attempt to insert observers into custom module. Later convert would ignore it as it is not in the swap_list. This is a problem in the example below (slightly modified) when observers cannot support the output format of the custom module.
```python
class customModule(torch.nn.Module):
  def __init__(self):
   ....

  def forward(self, x):
     # returns a list of tensors, not supported by observers
       return [x]


 class testModel(nn.Module):
    def __init__(self):
       super(testModel,self).__init__()
       self.conv1 = nn.Conv2d(...)
       self.myMod = customModule()
def forward(self,x):
   x= self.quant(x)
   x = self.conv1(x)
   x = self.dequant(x)
   x  = self.customModule(x)
   return(x)
```

raghuramank100 commented:
Control of observers in fake quant module
https://github.com/pytorch/pytorch/pull/27113

raghuramank100 commented:
Handle un-initialized values in Histogram observer
https://github.com/pytorch/pytorch/pull/27151

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -1070,6 +1071,15 @@ def test_observer_scriptable(self):
+        model.qconfig = torch.quantization.default_qconfig
+
+        model = prepare(model)
+        self.assertTrue(hasattr(model.fc1, 'qconfig'), "QConfig not propagating")
```
Please make the error message more informative, something like :"Expect Qconfig to propagate to linear module"

raghuramank100 commented on torch/quantization/default_mappings.py:
```diff
@@ -49,6 +49,7 @@
 }
 
 # List of modules to skip the qconfig propagation
+# TODO: Remove this after the white list is confirmed
```
Please remove prior to landing

raghuramank100 reviewed (approved):
Looks good, have a few comments.

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)

Commit:
Update on "[quantization] Replacing the skip_list with white_list in the qconfig propagation"


Differential Revision: [D17700548](https://our.internmc.facebook.com/intern/diff/D17700548)
